server/src/main/java/org/elasticsearch/common/geo/BoundingBox.java

```diff
@@ -0,0 +1,199 @@
+import org.elasticsearch.common.io.stream.Writeable;
+import org.elasticsearch.geometry.Geometry;
+import org.elasticsearch.geometry.Rectangle;
+import org.elasticsearch.geometry.ShapeType;
+import org.elasticsearch.geometry.utils.StandardValidator;
+import org.elasticsearch.geometry.utils.WellKnownText;
+import org.elasticsearch.xcontent.ParseField;
+import org.elasticsearch.xcontent.ToXContentFragment;
+import org.elasticsearch.xcontent.XContentBuilder;
+import org.elasticsearch.xcontent.XContentParser;
+
+import java.io.IOException;
+import java.text.ParseException;
+import java.util.Objects;
+
+/**
+ * A class representing a Bounding-Box for use by Geo and Cartesian queries and aggregations
+ * that deal with extents/rectangles representing rectangular areas of interest.
+ */
+public abstract class BoundingBox<T extends SpatialPoint> implements ToXContentFragment, Writeable {
+    static final ParseField TOP_RIGHT_FIELD = new ParseField("top_right");
+    static final ParseField BOTTOM_LEFT_FIELD = new ParseField("bottom_left");
+    static final ParseField TOP_FIELD = new ParseField("top");
+    static final ParseField BOTTOM_FIELD = new ParseField("bottom");
+    static final ParseField LEFT_FIELD = new ParseField("left");
+    static final ParseField RIGHT_FIELD = new ParseField("right");
+    static final ParseField WKT_FIELD = new ParseField("wkt");
+    public static final ParseField BOUNDS_FIELD = new ParseField("bounds");
+    public static final ParseField LAT_FIELD = new ParseField("lat");
+    public static final ParseField LON_FIELD = new ParseField("lon");
+    public static final ParseField TOP_LEFT_FIELD = new ParseField("top_left");
+    public static final ParseField BOTTOM_RIGHT_FIELD = new ParseField("bottom_right");
+    protected final T topLeft;
+    protected final T bottomRight;
+
+    public BoundingBox(T topLeft, T bottomRight) {
+        this.topLeft = topLeft;
+        this.bottomRight = bottomRight;
+    }
+
+    public boolean isUnbounded() {
+        return Double.isNaN(left()) || Double.isNaN(top()) || Double.isNaN(right()) || Double.isNaN(bottom());
+    }
+
+    public T topLeft() {
+        return topLeft;
+    }
+
+    public T bottomRight() {
+        return bottomRight;
+    }
+
+    public final double top() {
+        return topLeft.getY();
+    }
+
+    public final double bottom() {
+        return bottomRight.getY();
+    }
+
+    public final double left() {
+        return topLeft.getX();
+    }
+
+    public final double right() {
+        return bottomRight.getX();
+    }
+
+    @Override
+    public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+        builder.startObject();
+        toXContentFragment(builder, true);
+        builder.endObject();
+        return builder;
+    }
+
+    public abstract XContentBuilder toXContentFragment(XContentBuilder builder, boolean buildLatLonFields) throws IOException;
```

Comment:
`buildLatLonField` sounds very geo-specific. We should be able to get rid of this for now.